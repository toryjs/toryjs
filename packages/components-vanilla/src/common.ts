export type ReactComponent = string | React.ComponentType;
