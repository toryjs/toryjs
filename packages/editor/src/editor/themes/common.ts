export { Theme } from '@toryjs/ui';
