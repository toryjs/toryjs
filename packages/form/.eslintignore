node_modules
.fusebox
.cache
luis
public
dist
summary.json
snapshots.js
summary.ts
coverage