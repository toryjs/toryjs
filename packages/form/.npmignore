node_modules
src
wallaby.js
yarn-error.log
.eslintignore
.eslintrc.js
jest.config.js
.vscode
.github
jest.config.js
mocha.config.js