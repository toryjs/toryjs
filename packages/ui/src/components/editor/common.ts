import React from 'react';

export function calculatePosition(layout: string, e: React.DragEvent) {
  return layout === 'row' ? calculateVerticalPosition(e) : calculateHorizontalPosition(e);
}

export function calculateHorizontalPosition(e: React.DragEvent) {
  let element = e.currentTarget;
  let elementLeft = window.scrollX + element.getBoundingClientRect().left;
  let elementWidth = element.clientWidth;

  return e.pageX - elementLeft < elementWidth / 2 ? 'left' : 'right';
}

export function calculateVerticalPosition(e: React.DragEvent) {
  let element = e.currentTarget;
  let elementTop = window.scrollX + element.getBoundingClientRect().top;
  let elementHeight = element.clientHeight;
  let dragStart = e.pageY;

  // console.log(`${id} ${dragStart} - ${elementTop} < ${elementHeight} | ${dragStart - elementTop}`);

  return dragStart - elementTop < elementHeight / 2 ? 'top' : 'bottom';
}
