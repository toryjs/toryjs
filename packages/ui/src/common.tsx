import React from 'react';
import names from 'classnames';

import { interpolate } from '@tomino/toolbelt';
import { DataSet, FormElement, FormComponentProps, FormComponent } from '@toryjs/form';
import { BoundProp, BoundType } from '@toryjs/form';

import { css } from 'emotion';
import { toJS as mobxToJs } from 'mobx';
export { default as names } from 'classnames';

import { getPropValue } from './helpers';
import { ContextType } from './context';

export { Theme } from './editor/themes/common';
export { Context } from './context';
export {
  handle,
  getValue,
  isNullOrEmpty,
  prop,
  processControl,
  valueSource,
  createComponents
} from './helpers';
export { DynamicComponent as DynamicControl, paintProps } from './components/dynamic_component';

export const select = css`
  .react-select__control {
    border: 0px;
    border-radius: 0 !important;
    min-height: 23px;
  }

  .react-select__indicators {
    height: 23px;
  }

  .react-select__indicator {
    padding: 2px;
  }

  .react-select__value-container {
    height: 23px;

    > div {
      /* position: absolute; */
      margin: 0px;
      padding: 0px;
    }
  }

  .react-select__single-value {
    margin-top: 0px;
  }

  .react-select__menu {
    border-radius: 0px;
    margin-top: 2px;
  }

  label: select;
`;

export const fieldSet = css`
  /* name: fieldset */
  border: solid 1px #ddd;
  border-radius: 5px;
  background-color: #fafafa;

  legend {
    padding: 0px 6px;
    font-weight: bold;
  }
`;

export const breakingLabel = css`
  display: block;
  font-weight: bold;
  margin-top: 3px;
  margin-bottom: 3px;
`;

export const nonBreakingLabel = css`
  display: inline-block;
  font-weight: bold;
  padding-right: 6px;
  margin-top: 3px;
  margin-bottom: 3px;
`;

export const fullWidth = css`
  width: 100%;
`;

export const noPadding = css`
  padding: 0px !important;
`;

export const marginLess = css`
  margin: 0px !important;
`;

export const pointer = css`
  cursor: pointer !important;
  label: pointer;
`;

export const error = css`
  background: red;
  padding: 3px;
  border-radius: 3px;
  font-weight: bold;
  color: white;
`;

export type SignatureType = {
  comment: string;
  signature: string;
  verifiedState: 'Pending' | 'Verified' | 'Rejected';
  rejected?: boolean;
  uid: string;
  name: string;
  date: Date;
  setValue?(name: string, value: any): any;
};

export type SignatureHandlers = {
  validateForm(): boolean;
  verifySignature(owner: any, args: { uid: string; signature: string }): Promise<string>;
  sign(
    owner: DataSet,
    args: {
      source: string;
      reject: boolean;
      password: string;
      reason: string;
      submit: boolean;
    }
  ): Promise<SignatureType>;
  signatureFont(): string;
};

export type FormEditorProps = {
  formElement: DynamicForm.FormDataSet;
  formData: DataSet;
  handlers?: { [index: string]: any };
  readOnly: boolean;
};

export function parseFromOwner(str: string, owner: DataSet) {
  try {
    return interpolate(str, owner);
  } catch (ex) {
    return 'Error parsing string: ' + ex.message;
  }
}

export function sourceValue(prop: BoundProp) {
  return prop ? (prop as BoundType).source : null;
}

export function handlerValue(prop: BoundProp) {
  return prop ? (prop as BoundType).handler : null;
}

// let onClick: (props: FormComponentProps) => Function;
// export function registerOnClick(onClickHandler: typeof onClick) {
//   onClick = onClickHandler;
// }

export function root(formElement: FormElement) {
  let root = formElement;
  while (root.parent != null && (!root.pages || root.pages.length === 0)) {
    root = root.parent;
  }
  return root;
}

const extraProps = ['editorLabel'];

export function cleanProps(props: any) {
  const result: any = {};
  const js = toJS(props);
  for (let key of Object.keys(js)) {
    if (js[key] === '' || js[key] == null || js[key].length === 0 || extraProps.indexOf(key) >= 0) {
      continue;
    }
    result[key] = js[key];
  }
  return result;
}

export function tryInterpolate(text: string, owner: DataSet) {
  try {
    return interpolate(text, owner);
  } catch (ex) {
    return 'Error: ' + ex.message;
  }
}

export function useMergeState<T>(initialState: T): [T, (t: Partial<T>) => void] {
  const [state, setState] = React.useState(initialState);
  const setMergedState = (newState: any) =>
    setState((prevState: any) => Object.assign({}, prevState, newState));
  return [state, setMergedState];
}

// export function createComponents(props: FormComponentProps, className: string = null) {
//   if (!props.formElement.elements || props.formElement.elements.length === 0) {
//     return undefined;
//   }
//   return props.formElement.elements.map((e, i) => (
//     <React.Fragment key={i}>
//       {props.catalogue.createComponent(
//         { ...props, className: '' },
//         e,
//         className
//         // extraProps
//         // null,
//         // onClick ? onClick(props) : undefined
//       )}
//     </React.Fragment>
//   ));
// }

// type Options = {
//   drawLabel?: boolean;
//   className?: string;
//   onClick?: any
// }

export function createComponent1(
  props: FormComponentProps,
  formElement: FormElement,
  context: ContextType,
  className?: string,
  extraProps?: any
) {
  if (!formElement || formElement.control === 'EditorCell') {
    return null;
  }

  if (props.catalogue.components[formElement.control] === undefined) {
    return <div className={names(className, error)}>{formElement.control}</div>;
  }

  const hide = getPropValue(props, formElement, context, 'hidden', false);
  if (hide) {
    return null;
  }

  const controlDefinition = props.catalogue.components[formElement.control];
  const Control =
    (controlDefinition as FormComponent).Component ||
    (controlDefinition as React.ComponentType<any>);

  Control.displayName = 'CreateComponentWrapper';

  // TODO: Make this type safe by creating a render method in catalogue
  let renderedControl = (
    <Control
      owner={props.owner}
      className={className}
      handlers={props.handlers}
      catalogue={props.catalogue}
      formElement={formElement}
      extra={{ ...props.extra, ...extraProps }}
      dataProps={props.dataProps}
    />
  );

  return renderedControl;
}

export function toJS<T>(obj: T): T {
  return mobxToJs(obj);
}

export function addProviders(mainForm: FormElement, form: FormElement) {
  let formElement = form;

  // add all root providers from the man form
  if (form !== mainForm) {
    let model = toJS(mainForm);
    let provider = model;

    while (
      provider.elements &&
      provider.elements.length &&
      provider.elements[0].control.indexOf('Provider') >= 0
    ) {
      provider = provider.elements[0];
    }

    if (provider != model) {
      provider.elements = [form];
      formElement = model;
    }
  }

  return formElement;
}

// export function props<T>(props: T, defaultProps?: Partial<T>): T {
//   return controlProps || (defaultProps as any) || {};
// }
