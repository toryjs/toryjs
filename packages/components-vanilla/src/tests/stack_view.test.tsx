import { JSONSchema, FormElement } from '@toryjs/form';

import { prepareComponent, testStandard, testEditor, create } from './common';

function elements(): FormElement[] {
  return [
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 30px; background-color: red;` } }
    },
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 50px; background-color: blue;` } }
    },
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 20px; background-color: green;` } }
    },
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 20px; background-color: green;` } }
    },
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 20px; background-color: green;` } }
    },
    {
      control: 'Text',
      props: { value: '1', css: { value: `width: 100px; height: 20px; background-color: green;` } }
    }
  ];
}

const formDefinition: FormElement = create.stack({
  elements: [
    {
      control: 'Text',
      props: { value: 'Empty', control: 'Text' }
    },
    {
      control: 'Stack',
      props: {
        gap: '12px',
        layout: 'column'
      },
      elements: []
    },
    {
      control: 'Text',
      props: { value: 'Row Layout', control: 'Text' }
    },
    {
      control: 'Stack',
      props: {
        layout: 'row',
        gap: '0px'
      },
      elements: elements()
    },
    {
      control: 'Stack',
      props: {
        layout: 'row',
        gap: '6px'
      },
      elements: elements()
    },
    {
      control: 'Text',
      props: {
        value: 'Column Layout',
        control: 'Text',
        row: 3,
        column: 1,
        css: { value: 'margin-top: 15px;' }
      }
    },
    {
      control: 'Stack',
      props: {
        layout: 'column',
        gap: '12px',
        control: 'Stack'
      },
      elements: elements()
    },
    {
      control: 'Stack',
      props: {
        layout: 'column',
        gap: '0px',
        row: 5,
        column: 1
      },
      elements: elements()
    }
  ]
});

const formData = {};
const schema: JSONSchema = { type: 'object', properties: {} };

describe('Stack', () => {
  // it('renders correctly', () => {
  //   const wrapper = renderer.create(component());
  //   expect(wrapper).toMatchSnapshot();
  // });

  it('renders standard', () => {
    testStandard(formDefinition, schema, formData);
  });

  it('renders editor', () => {
    testEditor(formDefinition, schema, formData);
  });

  return {
    component: prepareComponent(formDefinition, schema, formData, {
      addEditor: true,
      showToolBox: true
    })
  };
});
