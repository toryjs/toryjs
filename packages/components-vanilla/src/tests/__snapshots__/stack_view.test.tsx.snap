exports["Stack renders standard 1"] = "\n<div\n  className=\"ui segment\"\n>\n  <div\n    className=\"ui dividing header\"\n  >\n    Standard\n  </div>\n  <div\n    className=\"ui form   css-1mu481t-form\"\n    data-control=\"Form\"\n  >\n    <div\n      data-control=\"Stack\"\n    >\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          dangerouslySetInnerHTML={\n            Object {\n              \"__html\": \"Empty\",\n            }\n          }\n          data-control=\"Text\"\n        />\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          data-control=\"Stack\"\n        />\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          dangerouslySetInnerHTML={\n            Object {\n              \"__html\": \"Row Layout\",\n            }\n          }\n          data-control=\"Text\"\n        />\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          data-control=\"Stack\"\n        >\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-4mdf4o\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-dg8zng\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-7hpbpj-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n        </div>\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          data-control=\"Stack\"\n        >\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-4mdf4o\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-dg8zng\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-1i51mle-item-row\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n        </div>\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          className=\"css-1pvpgaq\"\n          dangerouslySetInnerHTML={\n            Object {\n              \"__html\": \"Column Layout\",\n            }\n          }\n          data-control=\"Text\"\n        />\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          data-control=\"Stack\"\n        >\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-4mdf4o\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-dg8zng\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n          <div\n            className=\"css-eoi0db-item-column\"\n          >\n            <div\n              className=\"css-1iwh9a1\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"1\",\n                }\n              }\n              data-control=\"Text\"\n            />\n          </div>\n        </div>\n      </div>\n      <div\n        className=\"css-eoi0db-item-column\"\n      >\n        <div\n          data-control=\"Stack\"\n        >\n          <div\n            className=\"css-4mdf4o\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n          <div\n            className=\"css-dg8zng\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n          <div\n            className=\"css-1iwh9a1\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n          <div\n            className=\"css-1iwh9a1\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n          <div\n            className=\"css-1iwh9a1\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n          <div\n            className=\"css-1iwh9a1\"\n            dangerouslySetInnerHTML={\n              Object {\n                \"__html\": \"1\",\n              }\n            }\n            data-control=\"Text\"\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n";

exports["Stack renders editor 1"] = "\n<div\n  className=\"ui segment\"\n>\n  <div\n    className=\"ui dividing header\"\n  >\n    Editor\n  </div>\n  <div\n    style={\n      Object {\n        \"display\": \"flex\",\n      }\n    }\n  >\n    <div\n      style={\n        Object {\n          \"flex\": \"1\",\n        }\n      }\n    >\n      <div\n        className=\"ui form   css-1mu481t-form\"\n        data-control=\"Form\"\n        id=\"33\"\n      >\n        <div\n          className=\"css-1q7g8br\"\n          data-control=\"Stack\"\n          data-editor=\"isEditor\"\n          id=\"32\"\n          layout=\"column\"\n          onClick={[Function]}\n          onDragEnd={[Function]}\n          onDragLeave={[Function]}\n          onDragOver={[Function]}\n          onDrop={[Function]}\n          onMouseOut={[Function]}\n          onMouseOver={[Function]}\n        >\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"Empty\",\n                }\n              }\n              data-control=\"Text\"\n              id=\"0\"\n              onClick={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            />\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1q7g8br\"\n              data-control=\"Stack\"\n              data-editor=\"isEditor\"\n              id=\"1\"\n              layout=\"column\"\n              onClick={[Function]}\n              onDragEnd={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDrop={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            />\n            <div\n              className=\"css-1so7s8l-cell cellContent empty css-itj7i2-item-column\"\n              data-id=\"0\"\n              draggable={false}\n              onDragEnter={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDragStart={[Function]}\n              onDrop={[Function]}\n              onMouseOver={[Function]}\n            >\n               \n            </div>\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"Row Layout\",\n                }\n              }\n              data-control=\"Text\"\n              id=\"2\"\n              onClick={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            />\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1q7g8br\"\n              data-control=\"Stack\"\n              data-editor=\"isEditor\"\n              id=\"9\"\n              layout=\"row\"\n              onClick={[Function]}\n              onDragEnd={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDrop={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            >\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-4mdf4o\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"3\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-dg8zng\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"4\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"5\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"6\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"7\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-p1kag8-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"8\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n            </div>\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1q7g8br\"\n              data-control=\"Stack\"\n              data-editor=\"isEditor\"\n              id=\"16\"\n              layout=\"row\"\n              onClick={[Function]}\n              onDragEnd={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDrop={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            >\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-4mdf4o\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"10\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-dg8zng\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"11\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"12\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"13\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"14\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-1swo67l-item-row\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"15\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n            </div>\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1pvpgaq\"\n              dangerouslySetInnerHTML={\n                Object {\n                  \"__html\": \"Column Layout\",\n                }\n              }\n              data-control=\"Text\"\n              id=\"17\"\n              onClick={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            />\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1q7g8br\"\n              data-control=\"Stack\"\n              data-editor=\"isEditor\"\n              id=\"24\"\n              layout=\"column\"\n              onClick={[Function]}\n              onDragEnd={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDrop={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            >\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-4mdf4o\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"18\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-dg8zng\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"19\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"20\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"21\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"22\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n              <div\n                className=\"css-itj7i2-item-column\"\n              >\n                <div\n                  className=\"css-1iwh9a1\"\n                  dangerouslySetInnerHTML={\n                    Object {\n                      \"__html\": \"1\",\n                    }\n                  }\n                  data-control=\"Text\"\n                  id=\"23\"\n                  onClick={[Function]}\n                  onMouseOut={[Function]}\n                  onMouseOver={[Function]}\n                />\n              </div>\n            </div>\n          </div>\n          <div\n            className=\"css-itj7i2-item-column\"\n          >\n            <div\n              className=\"css-1q7g8br\"\n              data-control=\"Stack\"\n              data-editor=\"isEditor\"\n              id=\"31\"\n              layout=\"column\"\n              onClick={[Function]}\n              onDragEnd={[Function]}\n              onDragLeave={[Function]}\n              onDragOver={[Function]}\n              onDrop={[Function]}\n              onMouseOut={[Function]}\n              onMouseOver={[Function]}\n            >\n              <div\n                className=\"css-4mdf4o\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"25\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n              <div\n                className=\"css-dg8zng\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"26\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n              <div\n                className=\"css-1iwh9a1\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"27\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n              <div\n                className=\"css-1iwh9a1\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"28\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n              <div\n                className=\"css-1iwh9a1\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"29\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n              <div\n                className=\"css-1iwh9a1\"\n                dangerouslySetInnerHTML={\n                  Object {\n                    \"__html\": \"1\",\n                  }\n                }\n                data-control=\"Text\"\n                id=\"30\"\n                onClick={[Function]}\n                onMouseOut={[Function]}\n                onMouseOver={[Function]}\n              />\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n";

